go/vt/vtgate/planbuilder/route.go

```diff
@@ -440,15 +440,28 @@ func (rb *route) JoinCanMerge(pb *primitiveBuilder, rrb *route, ajoin *sqlparser
 		if where == nil {
 			return true
 		}
-		hasRuntimeRoutingPredicates := false
+		var tableWithRoutingPredicates sqlparser.TableNames
 		_ = sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
 			col, ok := node.(*sqlparser.ColName)
 			if ok {
-				hasRuntimeRoutingPredicates = hasRuntimeRoutingPredicates || isTableNameCol(col) || isDbNameCol(col)
+				hasRuntimeRoutingPredicates := isTableNameCol(col) || isDbNameCol(col)
+				if hasRuntimeRoutingPredicates {
+					for _, table := range tableWithRoutingPredicates {
+						if table.String() == col.Qualifier.String() {
+							return true, nil
+						}
+					}
+					tableWithRoutingPredicates = append(tableWithRoutingPredicates, col.Qualifier)
+				}
 			}
-			return !hasRuntimeRoutingPredicates, nil
+			return true, nil
 		}, where)
-		return !hasRuntimeRoutingPredicates
+		// Routes can be merged if only tables from information schema are accessed and only 1 table is used in the predicates that are used for routing
+		// TODO :- Even if more table are present in the routing, we can merge if they agree
+		if rb.ContainsOnlyInformationSchema() && rrb.ContainsOnlyInformationSchema() && len(tableWithRoutingPredicates) <= 1 {
+			return true
+		}
```

Comment:
The check for ContainsOnlyInformationSchema is redundant now, since we already check the opcode on line 437.